docs/controllers/optimization.md

```diff
@@ -0,0 +1,224 @@
+
+### Repeatedly Triggering Yourself
+
+AKA the problem that you will most easily run into; a **reconciler that modifies the status** (say) of its main [[object]] will cause a change in that object that is picked up by the [Controller]'s watcher loop, and will be **fed back into the reconciler**.
+
+This is normally not a problem, because **if** your status patch that causes this change is **idempotent**, it will **only happen once**. The problem is when you start putting non-deterministic values inside the the `.status` resource (e.g. timestamps rather than hashes).
+
+In such cases, the **controller will spin forever** on such objects.
+
+!!! note "Detecting spinlocks"
+
+    Spinlocks are usually noticeable quickly by just running the controller locally and watching logs for one object, or having a plot graph on your reconciler rate ([[observability#what-metrics]]) in `grafana`. You should expect about 1 to 5+ reconciles every `1/(your requeue time)` growing with the number of affected objects and self-interaction.
+
+The two ways to avoid reconciler re-triggering are:
+
+1. don't patch in non-deterministic values to your object (breaks idempotency)
+2. filter out changes to irrelevant parts of your object
+
+These approaches are **both recommended** because they both have independent merits;
+
+1. idempotency is good, and it avoids the spin problem the "good practice way"
+2. early filtering can allow for more precise reconcile bypasses with less code complexity, and allow opt-in non-determinism (such as timestamp fields)
+
+Watch events can be filtered out early using **predicates** with [WatchStreamExt::predicate_filter], and passing on these pre-filtered streams to controllers:
+
+```rust
+let deploys: Api<Deployment> = Api::default_namespaced(client);
+let changed_deploys = watcher(deploys, watcher::Config::default())
+    .applied_objects()
+    .predicate_filter(predicates::generation);
+```
+
+!!! warning "Predicates are unstable"
+
+    Predicates are a new feature in some flux with the last change in 0.84. They require one of the `unstable-runtime` feature flags.
+
+## Impossible Optimizations
+
+We leave a list of **currently impossible** (or technically infeasible) controller optimisation problems here and link to issues for transparency and in the hope that they will be tackled in the future.
+
+- [Sharing streams between **multiple controllers**](https://github.com/kube-rs/kube/issues/1080) (can only configure so far)
+- [Waiting for a store to be ready](https://github.com/kube-rs/kube/issues/1226)
+
+## Summary
+
+As a short summary, here are the main listed optimization and the effect you should expect to see from utilising them.
+
+| Optimization Type  | Target Reduction   |
+| ------------------ | ------------------ |
+| metadata_watcher   | IO + Memory        |
+| watcher selectors  | IO + Memory        |
+| watcher page size  | IO + Memory Spike  |
```

Comment:
Pagination is actually a tradeoff between IO and memory spikes, generally smaller pages mean lower memory usage but more IO.

Reply:
When you say IO tradeoff, are you referring to the IO overhead of having extra requests carry out the pagination? Isn't that just a few sets of request headers? I can include that if you think it's significant, but afaikt the actual data on the wire (which i feel is a "most of it") should be roughly the same :thinking:

Reply:
In terms of data, the overhead should be relatively small, yes (assuming that you're not doing something silly like `page_size(1)` for a metadata watcher). But there's also latency (fixed per-request overhead for things like authn/-z validation) and another synchronization point where the server must wait for the client to catch up (including confirmation!) before data streaming can continue. And in the worst case, each request could even mean doing a full TCP slow start and TLS handshake.

What it definitely won't do is reduce the amount of IO overhead.

Reply:
Ah, I see. Yes, it's probably wrong to highlight IO in the summary table. I was trying to highlight that it "spreads out the IO over time", making it less spiky, but that's not really important compared to the memory benefit.

Have reworded the table and changed up some page size paragraphs some more in https://github.com/kube-rs/website/pull/35/commits/d0bc8b9a9765796312d3cdc4ba07c73cbf05a0c6 to highlight it (but have avoided going into latency and IO here since it feels like the smaller problem to me).